readme.md
# LibPrice

A library for gathering an item's cost from various add-ons:

- [Master Merchant](https://www.esoui.com/downloads/info928-MasterMerchant.html), by Philgo68
- [Arkadius' Trade Tools](https://www.esoui.com/downloads/info1752-ArkadiusTradeTools.html), by Arkadius, Verbalinkontinenz
- [Tamriel Trade Centre](https://www.esoui.com/downloads/info1245-TamrielTradeCentre.html), by cyxui
- [Furniture Catalogue](https://www.esoui.com/downloads/info1617-FurnitureCatalogue.html), by manavortex

As well as some hard-coded data from

- The Crown Store
- Rolis Hlaalu, the Mastercraft Mediator
- Faustina Curio, the Achievement Mediator

## GitHub Repo

https://github.com/ziggr/ESO-LibPrice

## Example Code

```lua
function LibPrice_Example.SlashCommand()
    local self = LibPrice_Example

    for _,item_link in ipairs(EXAMPLE_ITEMS) do
                        -- Just tell me how much this thing costs.
        local gold   = LibPrice.ItemLinkToPriceGold(item_link)

                        -- Okay, I changed my mind. Tell me a little more:
                        -- where'd you get  this price from?
                        -- (additional return values from same API)
        gold, source_key, field_name
                     = LibPrice.ItemLinkToPriceGold(item_link)
        local header = self.FormatHeader(item_link, gold, source_key, field_name)
        d(header)

                        -- Give me all the data you can find, I'll figure out
                        -- what to do with it later.
        local result = LibPrice.ItemLinkToPriceData(item_link)
        LibPrice_Example.DumpTable(result)
    end
end
```
![example](doc/example.jpg)





doc/changelog.txt
[B]7.1.1 2021-09-14[/B]
[LIST]
[*] Master Merchant integration updated to MM 3.0. Thank you, Sharlikran!
[*] Update 31/ESO 7.1.0/API 101031/Walking Flame
[/LIST]

[B]6.2.1 2021-01-19[/B]
[LIST]
[*] Attempt to avoid nil error during FurnitureCatalogue lookup.
[*] API Bump to Update 28/6.2.1/API 100032/Markarth
[/LIST]

[B]5.0.2 2019-05-05[/B]
[LIST]
[*] ## IsLibrary: true so that Elsweyr can identify this as a library.
[/LIST]

[B]5.0.1 2019-05-05[/B]
[LIST]
[*] Version and API bump. No code change.
[/LIST]

[B]4.2.2 2018-12-01[/B]
[LIST]
[*] add crafting cost for craftable furnishings
[*] bugfix achievement furnishings
[*] bugfix Rolis Hlaalu furnishings
[*] add Music Box, Blood and Glory = 800 crowns
[/LIST]

[B]4.2.1 2018-11-29[/B]
[LIST]
[*] Initial release to support WritWorthy
[/LIST]



makepoll.pl
#! /usr/bin/perl

#$cmd = "make poll";
$cmd = "make put";

# Cheesy ls wrapper. Dir MUST end in /.
sub dirls($)
{
    my $dir = shift @_;
    my @r = `ls $dir`;
    foreach (@r)
    {
        chomp $_;
        $_ = $dir . $_;
    }
    return @r;
}

@files = ( 'Makefile' );
# push @files, dirls('src');
push @files, 'LibPrice.lua';
push @files, 'LibPrice_Internal.lua';
push @files, 'LibPrice_Example.lua';
print join("\n", @files) . "\n";

@prev_mtime = ();

$sleep_sec      =  2;
$dot_period_sec = 10;
$dot_sleep_sec  =  0;

sub mod_time()
{
    @mtime = ();
    foreach $file (@files)
    {
# print("FILE: $file .\n");
        @x = lstat $file;
        push @mtime, $x[9];
    }
    return @mtime;
}

for ( ; ; )
{
    @mtime = mod_time();

    if (join(" ", @mtime) ne join(" ", @prev_mtime))
    {
        clear_screen();
        print `$cmd 2>&1`;

        print STDERR "\n... waiting for changes ";
    }

    @prev_mtime = mod_time();

    sleep $sleep_sec;

    $dot_sleep_sec += $sleep_sec;
    if ($dot_period_sec <= ++$dot_sleep_sec)
    {
        $dot_sleep_sec = 0;
        print STDERR ".";
    }
}

sub clear_screen()
{
    my $clear      = "\e[2J";
    my $cursorhome = "\e[H";

    print "$clear$cursorhome";
}


LibPrice.lua
LibPrice = LibPrice or {}

-- Price ---------------------------------------------------------------------
--
-- Just tell me how much it costs.
--
-- Returns the first "suggested or average price in gold" that it can find.
-- Return nil if none found.
-- No crown or voucher prices returned.
--
-- Returns
--      gold price  (number, often a float)
--      source_key  (string "mm", "att", "ttc", others... )
--      field_name  (string "SuggestedPrice", "avgPrice", others... )
--
function LibPrice.ItemLinkToPriceGold(item_link, ...)
    local self   = LibPrice
    local field_names = { "SuggestedPrice", "Avg", "avgPrice", "npcVendor"}

                        -- If source list requested, then search only
                        -- the requested sources. If no source list requested,
                        -- search all sources.
    local requested_source_list = { ... }
    for _,source_key in ipairs(self.SourceList()) do
        if self.Enabled(source_key, requested_source_list) then
            local result = self.Price(source_key, item_link)
            if result then
                for _,field_name in ipairs(field_names) do
                    if result[field_name] then
                        return result[field_name], source_key, field_name
                    end
                end
            end
        end
    end
    return nil
end

-- All the data
--
-- input:
--  item_link
--  optional: list of sources to return:, any of:
--      "mm"
--      "att"
--      "furc"
--      "ttc"
--
-- Returns:
-- mm                   Master Merchant, by Philgo
--  avgPrice            https://www.esoui.com/downloads/info928-MasterMerchant.html
--  numSales            see MasterMerchant:tooltipStats() for what these
--  numDays             fields mean.
--  numItems
--  craftCost
--
-- att                  Arkadius Trade Tools, by Arkadius, Verbalinkontinenz
--  avgPrice            https://www.esoui.com/downloads/info1752-ArkadiusTradeTools.html
--  numDays             3-day or 90-day range used for this average?
--
-- furc                 Furniture Catalogue, by Manavortex
--  origin              https://www.esoui.com/downloads/info1617-FurnitureCatalogue.html
--  desc
--  currency_type
--  currency_ct
--  ingredient_list
--      ingr_ct
--      ingr_gold_ea
--      ingr_name
--      ingr_link
--      ingr_gold_sorce_key
--      ingr_gold_field_name
--
-- ttc                  Tamriel Trade Centre, by cyxui
--  Avg                 https://www.esoui.com/downloads/info1245-TamrielTradeCentre.html
--  Max                 see TamrielTradeCentre_PriceInfo:New() and
--  Min                 TamrielTradeCentrePrice:GetPriceInfo() for what these
--  EntryCount          fields mean.
--  AmountCount
--  SuggestedPrice
--
-- crown                Crown store: just a few items that Furniture Catalogue
--  crowns              lacked when Zig wrote this library's precursor in 2017.
--
-- rolis                Rolis Hlaalu, MasterCraft Mediator, and Faustina Curio,
--  vouchers            Achievement Mediator.
--
-- npc                  Sell to any NPC vendor for gold.
--  npcVendor
--
-- Not getting the price data you expect? Modify your item_link, perhaps
-- simplify some of those unimportant numbers. What does "simplify" and
-- "unimportant" mean here? Varies depending on item. Item links are...
-- enigmatic. See the [UESP page](https://en.uesp.net/wiki/Online:Item_Link)
-- for _some_ explanation.
--
function LibPrice.ItemLinkToPriceData(item_link, ... )
    local self   = LibPrice
    local result = {}
                        -- If source list requested, then search only
                        -- the requested sources. If no source list requested,
                        -- search all sources.
    local requested_source_list = { ... }
    for _,source_key in ipairs(self.SourceList()) do
        if self.Enabled(source_key, requested_source_list) then
            result[source_key] = self.Price(source_key, item_link)
        end
    end
    return result
end



LibPrice_Example.lua
LibPrice_Example = LibPrice_Example or {}

SLASH_COMMANDS["/example"] = function() LibPrice_Example.SlashCommand() end

local EXAMPLE_ITEMS = {
                    -- essence of health
      "|H1:item:54339:308:50:0:0:0:0:0:0:0:0:0:0:0:0:36:1:0:0:0:65536|h|h"
                    -- Crown Tri-Restoration Potion
    , "|H1:item:64710:123:1:0:0:0:0:0:0:0:0:0:0:0:1:36:0:1:0:0:0|h|h"
                    -- Recipe: Raven Rock Baked Ash Yams
    , "|H0:item:56970:4:1:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0|h|h"
                    -- Necklace of Willpower
    , "|H1:item:69278:363:50:0:0:0:0:0:0:0:0:0:0:0:257:24:0:1:0:0:0|h|h"
                    -- Trees, Paired Evergreens
    , "|H1:item:120550:3:1:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0|h|h"
                    -- Varla Stone
    , "|H1:item:134465:5:1:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0|h|h"
                    -- Tempering Alloy
    , "|H1:item:54173:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0|h|h"
                    -- Platinum Necklace, intricate
    , "|H0:item:138797:307:50:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0|h|h"
                    -- Jewelry station Naga Shaman
    , "|H1:item:145501:124:1:0:0:0:0:0:0:0:0:0:0:0:1:0:0:1:0:0:0|h|h"
                    -- Mundus Stone: The Warrior
    , "|H1:item:125453:6:1:0:0:0:0:0:0:0:0:0:0:0:65:0:0:1:0:0:0|h|h"
                    -- Transmute Station
    , "|H1:item:133576:6:1:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0|h|h"
    }

function LibPrice_Example.SlashCommand()
    local self = LibPrice_Example

    for _,item_link in ipairs(EXAMPLE_ITEMS) do
                        -- Just tell me how much this thing costs.
        local gold   = LibPrice.ItemLinkToPriceGold(item_link)

                        -- Okay, I changed my mind. Tell me a little more:
                        -- where'd you get  this price from?
                        -- (additional return values from same API)
        gold, source_key, field_name
                     = LibPrice.ItemLinkToPriceGold(item_link)
        local header = self.FormatHeader(item_link, gold, source_key, field_name)
        d(header)

                        -- Give me all the data you can find, I'll figure out
                        -- what to do with it later.
        local result = LibPrice.ItemLinkToPriceData(item_link)
        LibPrice_Example.DumpTable(result)
    end
end

function LibPrice_Example.FormatHeader(item_link, gold, source_key, field_name)
    local grey     = "|c999999"
    local white    = "|cFFFFFF"

    if not gold then
        local msg = string.format(
             "%s  "..grey.."price:unknown"
            , item_link )
        return msg
    end

    local gold_str = LibPrice_Example.GoldString(gold)
    local msg = string.format(
         "%s  "..grey.."price:"..white.."%s "..grey.." from %s.%s"
        , item_link
        , gold_str
        , source_key
        , field_name
        )
    return msg
end

function LibPrice_Example.GoldString(n)
    if not n then return tostring(n) end
    return ZO_LocalizeDecimalNumber(n)
            .." |t16:16:EsoUI/Art/currency/currency_gold.dds|t"
end

                        -- A recursive table dumper that does a marginally better
                        -- job of indenting than d()
function LibPrice_Example.DumpTable(t, indent_ct)
    local grey  = "|c999999"
    local white = "|cFFFFFF"
    local red   = "|cFF6666"
    if not t then
        d(red..tostring(t))
        return
    end
    indent_ct = indent_ct or 1
    local indent = string.format(".%"..(indent_ct*4).."."..(indent_ct*4).."s","")
                        -- Infinite recursion blocker
    if 4 < indent_ct then
        d(grey..indent.."...")
        return
    end
                        -- Can we squish down to single line?
    local has_sub_tables = false
    for k,v in pairs(t) do
        if type(v) == "table"
            and k ~= "points"   -- Omit MM's huge "points" history table.
            then
            has_sub_tables = true
            break
        end
    end
    if not has_sub_tables then
        local line = ""
        local sorted = LibPrice_Example.SortedKeys(t)
        for _,k in ipairs(sorted) do
            local v = t[k]
            if type(v) == "table" then
                        -- omit hex addr from tables that we choose not to
                        -- recurse into.
                v="table"
            end
            line = line .. string.format(grey.."%s:"..white.."%s  ",tostring(k),tostring(v))
        end
        d(indent..line)
    else
                        -- Gonna have to recurse, so print each key/value pair
                        -- on its own line.
        local sorted = LibPrice_Example.SortedKeys(t)
        for _,k in ipairs(sorted) do
            local v = t[k]
            local vv = tostring(v)
            if type(v) == "table" then vv = "table" end -- Omit useless hex addresses.
            d(string.format(grey..indent.."%-4s: "..white.."%s", tostring(k), vv))
            if type(v) == "table"
                and k ~= "points"       -- Omit MM's huge "points" history table.
                then
                LibPrice_Example.DumpTable(v,1+indent_ct)
            end
        end
    end
end
                        -- To provide more stable output.
function LibPrice_Example.SortedKeys(t)
    local r = {}
    for k,_ in pairs(t) do
        table.insert(r,k)
    end
    table.sort(r)
    return r
end


LibPrice_Internal.lua
LibPrice = LibPrice or {}

                        -- How long of a date range to pass to Master Merchant
                        -- and Arkadius Trade Tools.
LibPrice.day_ct_short =  3      -- ATT only
LibPrice.day_ct_long  = 90      -- MM and ATT

                        -- source keys
LibPrice.MM    = "mm"
LibPrice.ATT   = "att"
LibPrice.FURC  = "furc"
LibPrice.TTC   = "ttc"
LibPrice.CROWN = "crown"
LibPrice.ROLIS = "rolis"
LibPrice.NPC   = "npc"


local function Info(msg,...)
    d("|c999999LibPrice: "..string.format(msg,...))
end

local function Error(msg,...)
    d("|cFF6666LibPrice: "..string.format(msg,...))
end

-- Dispatch ------------------------------------------------------------------

function LibPrice.SourceList()
    if not LibPrice.SOURCE_LIST then
        LibPrice.SOURCE_LIST = { LibPrice.MM
                               , LibPrice.ATT
                               , LibPrice.FURC
                               , LibPrice.TTC
                               , LibPrice.CROWN
                               , LibPrice.ROLIS
                               , LibPrice.NPC
                               }
    end
    return LibPrice.SOURCE_LIST
end

function LibPrice.Price(source_key, item_link)
    if not source_key then return nil end
    if not item_link then return nil end
    local self = LibPrice
    if not self.DISPATCH then
        self.DISPATCH = {
          [self.MM   ] = { self.MMPrice    , self.CanMMPrice    }
        , [self.ATT  ] = { self.ATTPrice   , self.CanATTPrice   }
        , [self.FURC ] = { self.FurCPrice  , self.CanFurCPrice  }
        , [self.TTC  ] = { self.TTCPrice   , self.CanTTCPrice   }
        , [self.CROWN] = { self.CrownPrice }
        , [self.ROLIS] = { self.RolisPrice }
        , [self.NPC  ] = { self.NPCPrice   }
        }
    end
    if not (source_key and self.DISPATCH[source_key]) then
        Error("unknown source key:%s", tostring(source_key))
        return nil
    end
    if          self.DISPATCH[source_key][2]
        and not self.DISPATCH[source_key][2]() then
                        -- Requested source not installed/enabled.
        return nil
    end
    local cached = self.GetCachedPrice(source_key, item_link)
    if cached then
        -- Info("cached %s %s", item_link, source_key)
        return cached
    end
    local got = self.DISPATCH[source_key][1](item_link)
    if not got then
        -- Info("%s %s returned nil", item_link, source_key)
        return nil
    end
    self.SetCachedPrice(source_key, item_link, got)
    return got
end

                        -- If the caller requested a specific list of  sources,
                        -- then return true only if key is in that list.
                        --
                        -- If caller did not specify sources, then return true
                        -- for all keys.
function LibPrice.Enabled(key, source_list)
    if #source_list == 0 then return true end
    for _,k in ipairs(source_list) do
        if k == key then return true end
    end
    return false
end

-- Master Merchant ------------------------------------- Philgo, Sharlikran --

function LibPrice.CanMMPrice()
    return MasterMerchant and true
end

function LibPrice.MMPrice(item_link)
    if not (MasterMerchant and MasterMerchant.isInitialized) then return nil end
    if not item_link then return nil end
    local mm = MasterMerchant:itemStats(item_link, false)
    if not (mm and mm.avgPrice and 0 < mm.avgPrice) then
        return nil
    end
    return mm
end

-- Arkadius Trade Tools ----------------------- Arkadius, Verbalinkontinenz --

function LibPrice.CanATTPrice()
    return      ArkadiusTradeTools
            and ArkadiusTradeTools.Modules
            and ArkadiusTradeTools.Modules.Sales
            and ArkadiusTradeTools.Modules.Sales.addMenuItems
            and true
end

function LibPrice.ATTPrice(item_link)
    local self = LibPrice
                        -- ATT initializes its sales data at
                        -- EVENT_PLAYER_ACTIVATED time, which comes after
                        -- EVENT_ADD_ON_LOADED time, when some add-ons
                        -- *cough*WritWorthy*cough* foolishly request price
                        -- data. So check some random internal "addMenuItems"
                        -- that gets set during ATT sales data initialization
                        -- and return nil if ATT is not quite ready.
    if  not(    ArkadiusTradeTools
            and ArkadiusTradeTools.Modules
            and ArkadiusTradeTools.Modules.Sales
            and ArkadiusTradeTools.Modules.Sales.addMenuItems
            )
        then
        return nil
    end

    local day_secs = 24*60*60
    for _,day_ct in ipairs({ self.day_ct_short, self.day_ct_long }) do
        att = ArkadiusTradeTools.Modules.Sales:GetAveragePricePerItem(
                        item_link, GetTimeStamp() - (day_secs * day_ct))
        if att and 0 < att then
            return { avgPrice = att
                   , numDays  = day_ct
                   }
       end
    end
    return nil
end


-- Furniture Catalogue ----------------------------------------- Manavortex --

function LibPrice.CanFurCPrice()
    return FurC and true
end

function LibPrice.FurCPrice(item_link)
    if not item_link then return nil end
    if not FurC then return nil end
    local self = LibPrice

    local item_id       = FurC.GetItemId(item_link)
    local recipe_array  = FurC.Find(item_link)
    if not recipe_array then return nil end
    local origin        = recipe_array.origin
    if not origin then return nil end

    local desc           = FurC.GetItemDescription(item_id, recipe_array)
    local currency_type  = nil
    local currency_ct    = nil
    local currency_notes = nil

    local func_table = {
          [FURC_CRAFTING     ] = self.From_FurC_Crafting           --  3
        , [FURC_VENDOR       ] = self.From_FurC_AchievementVendor  --  6
        , [FURC_PVP          ] = self.From_FurC_PVP                --  7
        , [FURC_CROWN        ] = self.From_FurC_Crown              --  8
        , [FURC_LUXURY       ] = self.From_FurC_Luxury             -- 10
        , [FURC_ROLIS        ] = self.From_FurC_Rolis              -- 12
        , [FURC_DROP         ] = self.From_FurC_Misc               -- 14
        , [FURC_JUSTICE      ] = self.From_FurC_Misc               -- 15

        -- These tables never have per-item price data.
        , [FURC_RUMOUR       ] = self.From_FurC_NoPrice            --  9
        , [FURC_FESTIVAL_DROP] = self.From_FurC_NoPrice            -- 18
        }
    local func = func_table[origin] or self.From_FurC_Misc
    if func then  currency_type, currency_ct, currency_notes, ingredient_list
            = func(item_link, recipe_array)
    end

    local o = { origin          = origin
              , desc            = desc
              , currency_type   = currency_type
              , currency_ct     = currency_ct
              , notes           = currency_notes
              , ingredient_list = ingredient_list
              }
    return o
end


LibPrice.CURRENCY_TYPE_GOLD            = "gold"
LibPrice.CURRENCY_TYPE_WRIT_VOUCHERS   = "vouchers"
LibPrice.CURRENCY_TYPE_ALLIANCE_POINTS = "ap"
LibPrice.CURRENCY_TYPE_CROWNS          = "crowns"


-- Looking for FURC_XXX recipe_array.version values?
-- -- versioning
-- FURC_HOMESTEAD            = 2
-- FURC_MORROWIND            = 3
-- FURC_REACH                = 4
-- FURC_CLOCKWORK            = 5
-- FURC_DRAGONS              = 6
-- FURC_ALTMER               = 7
-- FURC_WEREWOLF             = 8
-- FURC_SLAVES               = 9

-- Dig deep into the data model of Furniture Catalogue,
-- because there's no public API for this (Furniture Catalogue
-- was never intended to become a public database).
--
function LibPrice.From_FurC_Crafting(item_link, recipe_array)
    local self = LibPrice
    if not recipe_array.blueprint then return nil, nil end
    local total_ingr_gold = 0
    local ingr_list       = {}
    local notes           = "Crafting cost"
    local blueprint_link  = FurC.GetItemLink(recipe_array.blueprint)
    local ingredient_ct   = GetItemLinkRecipeNumIngredients(blueprint_link)
    for ingr_i = 1, ingredient_ct do
        local _, _, ct    = GetItemLinkRecipeIngredientInfo(blueprint_link, ingr_i )
        local ingr_link   = GetItemLinkRecipeIngredientItemLink(blueprint_link, ingr_i )
        local ingr_name   = GetItemLinkName(ingr_link)
                        -- Recurse into each ingredient to get its gold cost.
                        -- If calling code wants more details about each
                        -- ingredient's cost, call ItemLinkToPriceData() on
                        -- each ingredient. I'm not going to do here.
        local gold, source_key, field_name = self.ItemLinkToPriceGold(ingr_link)
        local ingr_row = { ingr_ct              = ct
                         , ingr_name            = ingr_name
                         , ingr_link            = ingr_link
                         , ingr_gold_ea         = gold
                         , ingr_gold_source_key = source_key
                         , ingr_gold_field_name = field_name
                         }
        table.insert(ingr_list, ingr_row)
        if gold then
            total_ingr_gold = total_ingr_gold + ct * gold
        else
            notes = "Partial crafting cost, missing some ingredient costs."
        end
    end
    return self.CURRENCY_TYPE_GOLD, total_ingr_gold, notes, ingr_list
end

function LibPrice.From_FurC_Rolis(item_link, recipe_array)
    local self = LibPrice
    local item_id       = FurC.GetItemId(item_link)
    local seller_list   = { FurC.Rolis, FurC.Faustina }
    local seller_names  = { "Rolis", "Faustina" }
    local version_data  = nil
    for i, seller in ipairs(seller_list) do
        version_data = seller[recipe_array.version]
        if version_data and version_data[item_id] then break end
    end
    if not (version_data and version_data[item_id]) then return nil end
    local ct = version_data[item_id]
    return self.CURRENCY_TYPE_WRIT_VOUCHERS, ct, seller_names[i]
end

function LibPrice.From_FurC_Luxury(item_link, recipe_array)
    local self = LibPrice
    local version_data = FurC.LuxuryFurnisher[recipe_array.version]
    if not version_data then return nil end
    local item_id   = FurC.GetItemId(item_link)
    local item_data = version_data[item_id]
    if not item_data then return nil end
    return self.CURRENCY_TYPE_GOLD, item_data.itemPrice, "luxury vendor"
end

function LibPrice.From_FurC_AchievementVendor(item_link, recipe_array)
    local self = LibPrice
    local item_id      = FurC.GetItemId(item_link)
    local version_data = FurC.AchievementVendors[recipe_array.version]
    if not version_data then return nil end
    local entry = nil
    local notes = nil
    for location_name, zone_data in pairs(version_data) do
        for vendor_name, vendor_data in pairs(zone_data) do
            entry = vendor_data[item_id]
            if entry then
                notes = vendor_name .. " in " .. location_name
                return self.CURRENCY_TYPE_GOLD, entry.itemPrice, notes
            end
        end
    end
    return nil
end

function LibPrice.From_FurC_Generic(item_link, recipe_array, currency_type)
    local self = LibPrice
    local item_id      = FurC.GetItemId(item_link)
    local version_data = FurC.MiscItemSources[recipe_array.version]
    if not version_data then return nil, nil, nil end
    local origin_data  = version_data[recipe_array.origin]
    if not origin_data then return nil, nil, nil end
    local entry = origin_data[item_id]
    if type(entry) == "number" then
        return currency_type, entry, nil
    elseif type(entry) == "string" then
        local n = string.match(entry, "%d+")
        if n and tonumber(n) then
            return currency_type, tonumber(n), nil
        end
    elseif type(entry) == "table" then
        if entry.itemPrice then
            return currency_type, entry.itemPrice, nil
        end
    end
    return nil
end

function LibPrice.From_FurC_Crown(item_link, recipe_array)
    local self = LibPrice
    return self.From_FurC_Generic(item_link, recipe_array, self.CURRENCY_TYPE_CROWNS)
end

function LibPrice.From_FurC_Misc(item_link, recipe_array)
    local self = LibPrice
    return self.From_FurC_Generic(item_link, recipe_array, self.CURRENCY_TYPE_GOLD)
end

-- Rumor table and others lack any per-item details.
-- No point in wasting time in the Misc tables.
function LibPrice.From_FurC_NoPrice(item_link, recipe_array)
    return nil
end

function LibPrice.From_FurC_PVP(item_link, recipe_array, currency_type)
    local self = LibPrice
    local item_id      = FurC.GetItemId(item_link)
    local version_data = FurC.PVP[recipe_array.version]
    if not version_data then return nil, nil, nil end
    local entry = nil
    for vendor_name, vendor_data in pairs(version_data) do
        for location_name, location_data in pairs(vendor_data) do
            entry = location_data[item_id]
            if entry then break end
        end
    end
    if not entry then return nil end

    local notes = (vendor_name or "?") .. " in " .. (location_name or "?")
    return self.CURRENCY_TYPE_ALLIANCE_POINTS, entry.itemPrice, notes
end


-- Tamriel Trade Centre --------------------------------------------- cyxui --

function LibPrice.CanTTCPrice()
    return TamrielTradeCentrePrice and true
end

function LibPrice.TTCPrice(item_link)
    if not TamrielTradeCentrePrice then return nil end
    return TamrielTradeCentrePrice:GetPriceInfo(item_link)
end


-- Crown Store ------------------------------------------------------ ziggr --
--
-- Furniture Catalogue catches MOST crown store items. Here's a few more
-- that FurC did not list (at least way back in 2017 when Zig originally
-- wrote the precursor to this library.)

function LibPrice.CrownPrice(item_link)
    if not item_link then return nil end
    local self = LibPrice
    if not self.CASH then
        self.CASH = {
          ["The Apprentice"                         ] = { crowns   = 4000 }
        , ["The Atronach"                           ] = { crowns   = 4000 }
        , ["The Tower"                              ] = { crowns   = 4000 }
        , ["The Thief"                              ] = { crowns   = 4000 }
        , ["The Serpent"                            ] = { crowns   = 4000 }
        , ["The Ritual"                             ] = { crowns   = 4000 }
        , ["The Mage"                               ] = { crowns   = 4000 }
        , ["The Lady"                               ] = { crowns   = 4000 }
        , ["The Lord"                               ] = { crowns   = 4000 }
        , ["The Warrior"                            ] = { crowns   = 4000 }
        , ["The Lover"                              ] = { crowns   = 4000 }
        , ["The Steed"                              ] = { crowns   = 4000 }
        , ["The Shadow"                             ] = { crowns   = 4000 }
           --torage Coffer, Fortified"              ] = { crowns =  100 } one of these was free for reaching level 1
        , ["Storage Coffer, Secure"                 ] = { crowns =  100 }
        , ["Storage Coffer, Sturdy"                 ] = { crowns =  100 }
        , ["Storage Coffer, Oaken"                  ] = { crowns =  100 }
        , ["Storage Chest, Fortified"               ] = { crowns =  200 }
        , ["Storage Chest, Oaken"                   ] = { crowns =  200 }
        , ["Storage Chest, Secure"                  ] = { crowns =  200 }
        , ["Storage Chest, Sturdy"                  ] = { crowns =  200 }

        , ["Nuzhimeh the Merchant"                  ] = { crowns = 5000 }
        --["Pirharri the Smuggler"                  ] = ) -- Free for completing Thieves Guild quest lin
        , ["Tythis Andromo, the Banker"             ] = { crowns = 5000 }

        , ["Music Box, Blood and Glory"             ] = { crowns =  800 }

        , ["Imperial Pillar, Chipped"               ] = { crowns =  410 }
        , ["Imperial Pillar, Straight"              ] = { crowns =  410 }

        }
    end
                        -- EN English only for now, need to get proper links
                        -- and fix this.
    local name = GetItemLinkName(item_link)
    return self.CASH[name]
end

LibPrice.LINK_ATTUNABLE_BS = "|H1:item:119594:364:50:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:10000:0|h|h"
LibPrice.LINK_ATTUNABLE_CL = "|H1:item:119821:364:50:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:10000:0|h|h"
LibPrice.LINK_ATTUNABLE_WW = "|H1:item:119822:364:50:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:10000:0|h|h"
LibPrice.LINK_ATTUNABLE_JW = "|H1:item:137947:364:50:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:10000:0|h|h"


-- Rolis Hlaalu, Master Crafter Merchant ---------------------------- ziggr --

function LibPrice.RolisPrice(item_link)
    if not item_link then return nil end
    local self = LibPrice
    if not self.ROLIS_PRICE then
        self.ROLIS_PRICE = {
          [self.LINK_ATTUNABLE_BS] = { vouchers =  250 }
        , [self.LINK_ATTUNABLE_CL] = { vouchers =  250 }
        , [self.LINK_ATTUNABLE_WW] = { vouchers =  250 }
        , [self.LINK_ATTUNABLE_JW] = { vouchers =  250 }

        , ["Transmute Station"   ] = { vouchers = 1250 }
        }
    end
    local l = self.Unattune(item_link)
    local result = self.ROLIS_PRICE[l.item_link]

                        -- Fallback to EN English name for transmute
    if not result then
        local item_name = GetItemLinkName(item_link)
        result = self.ROLIS_PRICE[item_name]
    end
    return result
end

function LibPrice.Unattune(item_link)
    local self = LibPrice
                        -- Remove the "attuning" that makes these
                        -- crafting stations unique, preventing FurC
                        -- and MM from coming up with prices, AND
                        -- preventing us from combining them into a single
                        -- stack of count 44 instead of 44 unique items.
                        --
                        -- Note trailing space after each station name!
                        -- Required to avoid false match on non-attunable
                        -- stations.

                        -- EN English only for now. Fix later.
    local item_name = GetItemLinkName(item_link)

    if string.find(item_name, "Blacksmithing Station ") then
        return {
              item_name = "Attunable Blacksmithing Station"
            , item_link = self.LINK_ATTUNABLE_BS
            , furniture_data_id = 4050
            }
    elseif string.find(item_name, "Clothing Station ") then
        return {
              item_name = "Attunable Clothier Station"
            , item_link = self.LINK_ATTUNABLE_CL
            , furniture_data_id = 4051
            }
    elseif string.find(item_name, "Woodworking Station ") then
        return {
              item_name = "Attunable Woodworking Station"
            , item_link = self.LINK_ATTUNABLE_WW
            , furniture_data_id = 4052
            }
    elseif string.find(item_name, "Jewelry Crafting Station ") then
        return {
              item_name = "Attunable Jewelry Crafting Station"
            , item_link  = self.LINK_ATTUNABLE_JW
            , furniture_data_id = 4051
            }
    end
                        -- Nothing to unattune. Return original, unchanged.
    return { item_name = item_name
           , item_link = item_link
           }
end


-- NPC Vendor ----------------------------------------------------------------

function LibPrice.NPCPrice(item_link)
    local o = { GetItemLinkInfo(item_link) }
    if not (o and o[2] and 0 < o[2]) then return nil end
    return { npcVendor = o[2] }
end


-- Cache ---------------------------------------------------------------------
--
-- Calculating an "average price" from MM and ATT is an O(n) scan through
-- sales records. Expensive. Cache results for a few minutes to avoid crushing
-- the CPU with wasteful re-calculation.

LibPrice.cache              = nil
LibPrice.cache_reset_ts     = nil
LibPrice.CACHE_DUR_SECONDS  = 5 * 60

                        -- Allow cached data to expire after a few minutes.
function LibPrice.ResetCacheIfNecessary()
    local self = LibPrice
    local now_ts        = GetTimeStamp()
    self.cache_reset_ts = self.cache_reset_ts or now
    local prev_reset_ts = self.cache_reset_ts
    local ago_secs      = GetDiffBetweenTimeStamps(now_ts, prev_reset_ts)
    if self.CACHE_DUR_SECONDS < ago_secs then
        -- d("|cDD6666cache reset, ago_secs:"..tostring(ago_secs))
        self.cache = {}
        self.cache_reset_ts = now_ts
    else
        -- d("|c666666cache retained, ago_secs:"..tostring(ago_secs))
    end
end

function LibPrice.GetCachedPrice(source_key, item_link)
    LibPrice.ResetCacheIfNecessary()
    if not (  LibPrice.cache
            and LibPrice.cache[source_key]) then
        -- d("|cDDD666cache miss:"..item_link)
        return nil
    end
    -- d("|c66DD66cache hit:"..item_link)
    return LibPrice.cache[source_key][item_link]
end

function LibPrice.SetCachedPrice(source_key, item_link, value)
    LibPrice.cache                        = LibPrice.cache             or {}
    LibPrice.cache[source_key]            = LibPrice.cache[source_key] or {}
    LibPrice.cache[source_key][item_link] = value
end

